rosgraph_monitor/observer.py

```diff
@@ -62,12 +67,59 @@ def _stopped(self) -> bool:
         self._lock.release()
         return isSet
 
+
+class TopicObserver(Observer):
+    def __init__(self, name, topics,
+                loop_rate_hz=1,
+                qos_profile=QoSProfile(depth=5, history=HistoryPolicy.KEEP_LAST)):
+        super(TopicObserver, self).__init__(name, qos_profile, loop_rate_hz)
+
+        subscribers = []
+        for topic, topic_type in topics:
+            sub = Subscriber(self, topic_type, topic)
+            subscribers.append(sub)
+        topic_synchronizer = ApproximateTimeSynchronizer(
+            subscribers,
+            queue_size=5,
+            slop=5,
+            allow_headerless=True)
+        topic_synchronizer.registerCallback(self.generate_diagnostics)
+
+    # Example -- Every TopicObserver has to implement this function in a similar fashion
+    def calculate_attr(self, future, msg1, msg2):
+        res = msg1.data + msg2.data     # custom logic
+
+        status_msg = DiagnosticStatus()
+        status_msg.level = DiagnosticStatus.OK
+        status_msg.name = "Dummy Node"
+        status_msg.message = "QA status"
+
+        future.set_result(status_msg)
+
+    def publish_diagnostics(self, status_msg: DiagnosticStatus):
+        diag_msg = DiagnosticArray()
+
+        diag_msg.status = [status_msg]
+
+        diag_msg.header.stamp = self._clock.now().to_msg()
+        self._pub_diag.publish(diag_msg)
+
+    def generate_diagnostics(self, *msgs) -> t.List[DiagnosticStatus]:
+        future = Future()
+        future.add_done_callback(lambda future : self.publish_diagnostics(future.result()))
```

Comment:
can there be a timeout to receive messages? At the end of timeout a "no messages received" ERROR can be published